vslib/src/SwitchStateBaseMACsec.cpp

```diff
@@ -0,0 +1,831 @@
+    CHECK_STATUS(findPortByMACsecFlow(flow_id, portId));
+    CHECK_STATUS(findHostInterfaceInfoByPort(portId, macsecAttr.m_info));
+    macsecAttr.m_vethName = vs_get_veth_name(macsecAttr.m_info->m_name, portId);
+    macsecAttr.m_macsecName = SAI_VS_MACSEC_PREFIX + macsecAttr.m_vethName;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_AN, attrCount, attrList);
+
+    macsecAttr.m_an = attr->value.u8;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_ENCRYPTION_ENABLE, attrCount, attrList);
+
+    macsecAttr.m_encryptionEnable = attr->value.booldata;
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_SAK, attrCount, attrList);
+
+    macsecAttr.m_sak = sai_serialize_hex_binary(attr->value.macsecsak);
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_SAK_256_BITS, attrCount, attrList);
+
+    if (!attr->value.booldata)
+    {
+        macsecAttr.m_sak = macsecAttr.m_sak.substr(macsecAttr.m_sak.length() / 2,macsecAttr.m_sak.length() / 2);
+    }
+
+    SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_AUTH_KEY, attrCount, attrList);
+    macsecAttr.m_authKey = sai_serialize_hex_binary(attr->value.macsecauthkey);
+
+    if (macsecAttr.m_direction == SAI_MACSEC_DIRECTION_EGRESS)
+    {
+        SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_XPN, attrCount, attrList);
+    }
+    else
+    {
+        SAI_METADATA_GET_ATTR_BY_ID(attr, SAI_MACSEC_SA_ATTR_MINIMUM_XPN, attrCount, attrList);
+    }
+
+    macsecAttr.m_pn = attr->value.u64;
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchStateBase::loadMACsecAttr(
+    _In_ sai_object_type_t objectType,
+    _In_ sai_object_id_t objectId,
+    _In_ uint32_t attrCount,
+    _In_ const sai_attribute_t *attrList,
+    _Out_ MACsecAttr &macsecAttr)
+{
+    SWSS_LOG_ENTER();
+
+    if (attrList == nullptr || attrCount == 0)
+    {
+        SWSS_LOG_ERROR("Attribute list is empty");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    switch (objectType)
+    {
+    case SAI_OBJECT_TYPE_MACSEC_PORT:
+        return loadMACsecAttrFromMACsecPort(objectId, attrCount, attrList, macsecAttr);
+
+    case SAI_OBJECT_TYPE_MACSEC_SC:
+        return loadMACsecAttrFromMACsecSC(objectId, attrCount, attrList, macsecAttr);
+
+    case SAI_OBJECT_TYPE_MACSEC_SA:
+        return loadMACsecAttrFromMACsecSA(objectId, attrCount, attrList, macsecAttr);
+
+    default:
+        SWSS_LOG_ERROR("Wrong type %s", sai_serialize_object_type(objectType));
+
+        break;
+    }
+
+    return SAI_STATUS_FAILURE;
+}
+
+sai_status_t SwitchStateBase::loadMACsecAttr(
+    _In_ sai_object_type_t objectType,
+    _In_ sai_object_id_t objectId,
+    _Out_ MACsecAttr &macsecAttr)
+{
+    SWSS_LOG_ENTER();
+
+    std::vector<sai_attribute_t> attrs;
+
+    if (dumpObject(objectId, attrs))
+    {
+        if (loadMACsecAttr(objectType, objectId, static_cast<uint32_t>(attrs.size()), attrs.data(), macsecAttr) != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_DEBUG(
```

Comment:
error/warning ?

Reply:
I think it should be DEBUG. 
An example as follow:
If we want to create a MACsec SA, we need to wait two conditions. The first one is MACsec SA ready and the second one is the ACL entry was set to MACsec flow.
This function `loadMACsecAttr` will try to check these two conditions. I don't think it means orchagent provides a wrong configuration if to load MACsec fails. Maybe it just means we need to wait another command from orchagent.

For the specific error, I recorded them in the concrete position.